Drop Legacy Cascading Dependencies From the Frontend

CodaFi opened the pull request:
It's been a long time coming.

CodaFi committed:
Remove TypeResolutionFlags::KnownNonCascadingDependency

CodaFi committed:
Remove UnqualifiedLookupFlags::KnownPrivate

CodaFi committed:
Remove NameLookupFlags::KnownPrivate

CodaFi committed:
Remove NLOptions::NL_Known*

CodaFi committed:
Remove DeclContext::isCascadingContextForLookup

CodaFi committed:
Remove evaluator::getScopeForAccessLevel

CodaFi committed:
Remove DependencyScope

CodaFi committed:
Remove Reference::cascades

CodaFi committed:
Remove unnecessary dependency sources

CodaFi committed:
Remove Cascading Computations from ASTScope

CodaFi committed:
Remove LegacyCascadingDependencies

CodaFi committed:
Drop Legacy Tests

CodaFi committed:
Remove appendHolderOfCascadingMembers

CodaFi committed:
Remove DependencyRecorder::Mode

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test source compatibility

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test source compatibility Debug

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
@swift-ci test source compatibility Debug

davidungar commented on include/swift/AST/ASTScope.h:
```diff
@@ -618,7 +610,6 @@ class Portion {
   getChildlessSourceRangeOf(const GenericTypeOrExtensionScope *scope,
                             bool omitAssertions) const = 0;
 
-  /// Returns isDone and isCascadingUse
```
What does it return now? Just isDone? Unless a function name gives the predicate, I think a comment can really help.

CodaFi replied:
I believe it returns `isDone` given that it calls `lookupMembersOf`, and that function no longer computes the cascading bit. I'll add this back once the testing infrastructure gives me the green light.

davidungar replied:
Thanks!

davidungar commented on include/swift/AST/EvaluatorDependencies.h:
```diff
@@ -32,80 +33,13 @@ namespace detail {
-
-// A \c DependencySource is currently defined to be a parent source file and
-// an associated dependency scope.
+// A \c DependencySource is currently defined to be a parent source file.
```
"parent"? Enclosing lexically? Or lookup parent?

CodaFi replied:
Now that we're in pure-private-dependencies-land, the source is always a primary. Going to update this.

davidungar replied:
Thanks!

davidungar commented on include/swift/AST/EvaluatorDependencies.h:
```diff
@@ -32,80 +33,13 @@ namespace detail {
-// an associated dependency scope.
+// A \c DependencySource is currently defined to be a parent source file.
 //
 // The \c SourceFile instance is an artifact of the current dependency system,
 // and should be scrapped if possible. It currently encodes the idea that
 // edges in the incremental dependency graph invalidate entire files instead
 // of individual contexts.
```
Love this comment!

davidungar commented on lib/AST/Evaluator.cpp:
```diff
@@ -381,8 +371,7 @@ void Evaluator::dumpDependenciesGraphviz() const {
 void evaluator::DependencyRecorder::realize(
     const DependencyCollector::Reference &ref) {
-  auto *source = getActiveDependencySourceOrNull();
-  assert(source && "cannot realize dependency without associated file!");
```
Is it worth keeping this assertion?

CodaFi replied:
I removed it since `NullablePtr::get` is going to fire its own assertion if we get this wrong at some point. I could leave it in so we get more diversified stack traces, but since it's an internal invariant if it fails something catastrophic really has gone wrong.

davidungar replied:
Makes sense. Thanks for explaining.

davidungar commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -105,11 +105,6 @@ namespace {
     using ResultsVector = SmallVector<LookupResultEntry, 4>;
     
   private:
-    struct ContextAndResolvedIsCascadingUse {
-      DeclContext *const DC;
-      const bool isCascadingUse;
-    };
-    
```
This is a huge PR to review, but it's so nice it's well worth it!

davidungar commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -119,34 +114,35 @@ namespace {
       /// Nontypes are formally members of the base type, i.e. the dynamic type
       /// of the activation record.
-      DeclContext *const dynamicContext;
+      const DeclContext *const dynamicContext;
```
Love those `const`s!

davidungar reviewed (approved):
Fantastic!

CodaFi commented:
@swift-ci test macOS

CodaFi commented:
[Saving a link to the pre-primary compiler perf run](https://ci.swift.org/job/swift-PR-compiler-performance-macOS/707/)

CodaFi committed:
[NFC] Only Register Primaries As Dependency Sources

Now that the top-level source file is the only dependency source that
matters, the only case that matters is when request evaluation enters
a primary file. For non-primaries, there will be no corresponding
swiftdeps file to emit references into, so we're just wasting time and
memory keeping track of anything that happens there.

This is only possible after we removed cascading dependencies because
unqualified lookups had to be charged to the files they originated in.
Now, we charge those lookups to the primary that initiated the request.

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci smoke test Windows

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
These numbers make no sense. No way this PR cuts 90% of wall time alone, we just failed an enormous amount of projects. 

@swift-ci test compiler performance

slavapestov commented on lib/AST/FrontendSourceFileDepGraphFactory.cpp:
```diff
@@ -580,7 +580,7 @@ class UsedDeclEnumerator {
     auto &Ctx = SF->getASTContext();
     std::unordered_set<std::string> holdersOfCascadingMembers;
```
This set is now always empty and an be removed

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi committed:
Resolve ASTScope Issues

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test source compatibility

CodaFi commented:
⛵
